### PR TITLE
Enable executable caching when type inference is disabled

### PR DESCRIPTION
**Pull request by DenisVieriu97**
Fixes perf regression in torchbench (Brings back changes from https://github.com/kulinseth/pytorch/pull/382)

### CODE REVIEW

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 1]
```diff
@@ -8,7 +8,36 @@
...
+    MPSGraph *mpsGraph = cachedGraph->graph();
+    MPSGraphExecutable* executable = cachedGraph->getExecultable();
+    if (!executable) {
+      NSMutableDictionary* shapes = [[NSMutableDictionary new] autorelease];
+      for (MPSGraphTensor* graphTensor in feeds) {
+        MPSGraphTensorData* graphTensorData = [feeds objectForKey:graphTensor];
+        shapes[graphTensor] = [[[MPSGraphShapedType alloc] initWithShape:nil dataType:graphTensorData.dataType] autorelease];
+      }
+      if (alphaTensor) {
+        shapes[alphaTensor] = [[[MPSGraphShapedType alloc] initWithShape:@[@1] dataType:alphaTensor.dataType] autorelease];
```
Comment (DenisVieriu97) [inline aten/src/ATen/native/mps/OperationUtils.mm:26]: This is needed for the alpha tensor ranked placeholder which is always 1. I will update that tensor to be unranked when type inference is disabled.

File: aten/src/ATen/mps/MPSStream.mm [hunk 2]
```diff
@@ -20,6 +20,7 @@ @interface MPSGraphExecutionDescriptor ()
   TORCH_CHECK(_stream.device_type() == DeviceType::MPS);
   _serialQueue = dispatch_queue_create("metal gpu stream", nullptr);
   _executionDescriptor = [MPSGraphExecutionDescriptor new];
+  _executableExecutionDescriptor = [MPSGraphExecutableExecutionDescriptor new];
```
Comment (razarmehr) [inline aten/src/ATen/mps/MPSStream.mm:23]: Release and nil in destructor? Otherwise leaks.

File: aten/src/ATen/mps/MPSStream.mm [hunk 3]
```diff
@@ -189,21 +190,36 @@ @interface MPSGraphExecutionDescriptor ()
...
+        }
+        
+        for (MPSGraphTensor *tensor in [executable targetTensors]) {
+          resultsArray[ouputIndex++] = results[tensor];
+        }
+
+        MPSGraphExecutableExecutionDescriptor *executionDescriptor =
+                        [[MPSGraphExecutableExecutionDescriptor new] autorelease];
+
+        executionDescriptor.completionHandler = ^(NSArray<MPSGraphTensorData *> * _Nonnull,
```
Comment (razarmehr) [inline aten/src/ATen/mps/MPSStream.mm:208]: This is not required (causes unnecessary overhead of calling an empty callback). 
If `executionDescriptor.completionHandler` is nil, the MPSGraph will not run it.

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 4]
```diff
@@ -8,7 +8,36 @@
 
 void runMPSGraph(MPSStream* mpsStream, MPSGraph* mpsGraph, NSDictionary* feeds,
                  NSDictionary* results, bool disableTypeInference) {
-  mpsStream->executeMPSGraph(mpsGraph, feeds, results, SyncType::COMMIT_ADAPTIVE, disableTypeInference);
+  mpsStream->executeMPSGraph(mpsGraph, feeds, results, SyncType::COMMIT_ADAPTIVE);
+}
+
+// this should be merged into runMPSGraph() with new arg "disableTypeInference" for executables
+void runMPSGraphExecutable(MPSStream *mpsStream, MPSCachedGraph *cachedGraph, NSDictionary *feeds, NSDictionary *results, MPSGraphTensor* alphaTensor) {
```
Comment (razarmehr) [inline aten/src/ATen/native/mps/OperationUtils.mm:15]: This is a generic function and Binary Op-specific arguments like alphaTensor shouldn't be in here.
As discussed, please make the alphaTensor unranked in its origin (i.e., Binary ops) when type inference is disabled).

**Review by razarmehr — CHANGES_REQUESTED** [review]

File: aten/src/ATen/mps/MPSStream.mm (hunk 3 above)
Comment (DenisVieriu97) [reply aten/src/ATen/mps/MPSStream.mm]: Thanks. Updated

File: aten/src/ATen/native/mps/OperationUtils.mm (hunk 4 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/OperationUtils.mm:15]: Thanks, updated.

Comment (razarmehr) [reply aten/src/ATen/native/mps/OperationUtils.mm:15]: You don't need to pass alphaTensor here. it should already be in the feeds.

**Review by razarmehr — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mps/OperationUtils.mm (hunk 4 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/OperationUtils.mm]: Yea, I wanted to get rid completely of the `runMPSGraphExecutable` function - and just have a generic function with `runMPSGraph`. I've updated that now

**Review by razarmehr — APPROVED** [review]
Looks good.

**Pull request merged**